src/contracts/MsVault.h

```diff
@@ -0,0 +1,708 @@
+                        output.vaultNames.set(locals.count, v.vaultName);
+                        locals.count++;
+                        break;
+                    }
+                }
+            }
+        }
+        output.numberOfVaults = locals.count;
+    _
+
+    PUBLIC_FUNCTION_WITH_LOCALS(getReleaseStatus)
+        if (input.vaultID >= MAX_VAULTS)
+        {
+            for (locals.i = 0; locals.i < MAX_OWNERS; locals.i++)
+            {
+                output.amounts.set(locals.i, 0);
+                output.destinations.set(locals.i, NULL_ID);
+            }
+            return;
+        }
+
+        Vault v = state.vaults.get(input.vaultID);
+        if (!v.isActive)
+        {
+            for (locals.i = 0; locals.i < MAX_OWNERS; locals.i++)
+            {
+                output.amounts.set(locals.i, 0);
+                output.destinations.set(locals.i, NULL_ID);
+            }
+            return;
+        }
+
+        for (locals.i = 0; locals.i < v.numberOfOwners; locals.i++)
+        {
+            output.amounts.set(locals.i, v.releaseAmounts.get(locals.i));
+            output.destinations.set(locals.i, v.releaseDestinations.get(locals.i));
+        }
+    _
+
+    PUBLIC_FUNCTION_WITH_LOCALS(getBalanceOf)
+        if (input.vaultID >= MAX_VAULTS)
+        {
+            output.balance = 0;
+            return;
+        }
+
+        locals.vault = state.vaults.get(input.vaultID);
+        if (!locals.vault.isActive)
+        {
+            output.balance = 0;
+            return;
+        }
+        output.balance = locals.vault.balance;
+    _
+
+    PUBLIC_FUNCTION_WITH_LOCALS(getVaultName)
+        if (input.vaultID >= MAX_VAULTS)
+        {
+            output.vaultName = NULL_ID;
+            return;
+        }
+
+        locals.vault = state.vaults.get(input.vaultID);
+        if (!locals.vault.isActive)
+        {
+            output.vaultName = NULL_ID;
+            return;
+        }
+        output.vaultName = locals.vault.vaultName;
+    _
+
+    PUBLIC_FUNCTION(getRevenueInfo)
+        output.numberOfActiveVaults = state.numberOfActiveVaults;
+        output.totalRevenue = state.totalRevenue;
+        output.totalDistributedToShareholders = state.totalDistributedToShareholders;
+    _
+
+    INITIALIZE
+        state.numberOfActiveVaults = 0;
+        state.totalRevenue = 0;
+        state.totalDistributedToShareholders = 0;
+    _
+
+    struct BEGIN_EPOCH_locals
+    {
+        uint64 i;
+        resetReleaseRequests_input rr_in;
+        resetReleaseRequests_output rr_out;
+        resetReleaseRequests_locals rr_locals;
+    };
+
+    BEGIN_EPOCH_WITH_LOCALS
+        for (locals.i = 0; locals.i < MAX_VAULTS; locals.i++)
+        {
+            Vault v = state.vaults.get(locals.i);
+            if (v.isActive)
+            {
+                locals.rr_in.vault = v;
+                resetReleaseRequests(qpi, state, locals.rr_in, locals.rr_out, locals.rr_locals);
```

Comment:
why should the release request not being kept over epochs?

Reply:
@krypdkat what do you think?
after rethinking it makes more sense to keen the requests. from users perspective it is best practice that they can send release requests at any time. may be adding a timeout, but for simplicity i would just keep it active until manual reset or acceptance.

Reply:
@J0ET0M  the main purpose was only trying to make a clean state at every epoch begin (easier debugging and calculation), nothing more.
So if the code and other logic are concrete, then reseting storage is not needed.